x-pack/solutions/search/plugins/search_connectors/server/task.ts

```diff
@@ -0,0 +1,196 @@
+) {
+  return ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+    return {
+      run: async () => {
+        try {
+          // We fetch some info even if license does not permit actual operations.
+          // This is done so that we could give a warning to the user only
+          // if they are actually using the feature.
+          logger.debug('Checking state of connectors and agentless policies');
+
+          // Fetch connectors
+          const nativeConnectors = await service.getNativeConnectors();
+
+          const policiesMetadata = await service.getConnectorPackagePolicies();
+
+          // Check license if any native connectors or agentless policies found
+          if (nativeConnectors.length > 0 || policiesMetadata.length > 0) {
+            const license = await licensingPluginStart.getLicense();
+
+            if (license.check('fleet', 'platinum').state !== 'valid') {
+              logger.warn(
+                'Current license is not compatible with agentless connectors. Please upgrade the license to at least platinum'
+              );
+              return;
+            }
+          }
+
+          // Deploy Policies
+          const connectorsWithoutPolicies = getConnectorsWithoutPolicies(
+            policiesMetadata,
+            nativeConnectors
+          );
+
+          let agentlessConnectorsDeployed = 0;
+          for (const connectorMetadata of connectorsWithoutPolicies) {
+            // We try-catch to still be able to deploy other connectors if some fail
+            try {
+              await service.deployConnector(connectorMetadata);
+
+              agentlessConnectorsDeployed += 1;
+            } catch (e) {
+              logger.warn(
+                `Error creating an agentless deployment for connector ${connectorMetadata.id}: ${e.message}`
+              );
+            }
+          }
+
+          // Delete policies
+          const policiesWithoutConnectors = getPoliciesWithoutConnectors(
+            policiesMetadata,
+            nativeConnectors
+          );
+          let agentlessConnectorsRemoved = 0;
+
+          for (const policyMetadata of policiesWithoutConnectors) {
+            // We try-catch to still be able to deploy other connectors if some fail
+            try {
+              await service.removeDeployment(policyMetadata.package_policy_id);
+
+              agentlessConnectorsRemoved += 1;
+            } catch (e) {
+              logger.warn(
+                `Error when deleting a package policy ${policyMetadata.package_policy_id}: ${e.message}`
+              );
+            }
+          }
+          return {
+            state: {},
+            schedule: SCHEDULE,
+          };
+        } catch (e) {
+          logger.warn(`Error executing agentless deployment sync task: ${e.message}`);
+          return {
+            state: {},
+            schedule: SCHEDULE,
+          };
+        }
+      },
+      cancel: async () => {
```

Comment:
Note that if you want to have the `cancel` actually stop the task from running, you'll have to do a bit more.  This function is invoked when TM decides the task needs to be cancelled (running longer than it's time limit).  The basic idea is you set a local indicating you've been cancelled, and then can check that in the `run()` method.  Example here: https://github.com/elastic/kibana/blob/0e13d86fc7b37c48011b9a1e601ae9f4e7d664d9/x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/entity_store/task/field_retention_enrichment_task.ts#L250-L279 - note that this code doesn't actually seem to use the `isCancelled()` local function they created - I think it did at one point, must have been removed in another PR ...

Reply:
Thanks, I see now - I copied original code from some other place.

The task we have is supposed to be very fast (obviously, depending on response times from Elasticsearch). At best it's 2 calls to Elasticsearch, at worst it's probably does 10s of calls (realistically, I expect 2 calls 99.99% of the time, and occasionally 4 calls).

IMO adding true cancellation is not gonna add a lot here, but I'm not sure yet. I will merge what is there for now and keep it in mind for when we need it :)